onnxruntime/python/tools/symbolic_shape_infer.py

```diff
@@ -796,10 +801,14 @@ def _infer_If(self, node):
                     vi.CopyFrom(subgraph.output[i_out])
                     vi.name = node.output[i_out]
                 else:
-                    assert all([
-                        d1 == d2 for d1, d2 in zip(vi.type.tensor_type.shape.dim,
-                                                   subgraph.output[i_out].type.tensor_type.shape.dim)
-                    ])
+                    for di, ds in enumerate(zip(vi.type.tensor_type.shape.dim,
+                                                subgraph.output[i_out].type.tensor_type.shape.dim)):
+                        # The two branches have different dimension in the output shapes
+                        # create a new symbolic dimension
+                        if ds[0] != ds[1]:
+                            new_dim = onnx.TensorShapeProto.Dimension()
+                            new_dim.dim_param = self._new_symbolic_dim_from_output(node, i_out, di)
+                            vi.type.tensor_type.shape.dim[di].CopyFrom(new_dim)
                 # pass on sympy data from subgraph, if cond is constant
                 if cond is not None and i_sub == (0 if cond > 0 else 1):
```

Comment:
"cond > 0" might change to "as_scaler(cond) > 0", otherwise error will occur when cond is a list.